File: Interview Problems/Parking Lot/Java Code/repository/PaymentRepository.java
package repository;

import domain.Payment;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

public class PaymentRepository {

    // A ticket may have multiple payment attempts, so store as List per ticketId
    private final Map<UUID, Payment> store = new HashMap<>();
    private final Map<UUID, List<UUID>> ticketToPayments = new HashMap<>();

    public void save(Payment payment) {
        store.put(payment.getId(), payment);
        ticketToPayments
            .computeIfAbsent(payment.getTicketId(), k -> new ArrayList<>())
            .add(payment.getId());
        System.out.println("[REPO] Payment saved: " + payment.getId() + " for ticket: " + payment.getTicketId());
    }

    public Optional<Payment> findById(UUID paymentId) {
        return Optional.ofNullable(store.get(paymentId));
    }

    /**
     * Returns the most recent payment attempt for a ticket.
     * Used during retry to check if a prior attempt succeeded (to avoid double-charging).
     */
    public Optional<Payment> findByTicketId(UUID ticketId) {
        List<UUID> paymentIds = ticketToPayments.get(ticketId);
        if (paymentIds == null || paymentIds.isEmpty()) return Optional.empty();
        // Return the last attempt
        UUID lastId = paymentIds.get(paymentIds.size() - 1);
        return Optional.ofNullable(store.get(lastId));
    }

    public List<Payment> findAllByTicketId(UUID ticketId) {
        List<UUID> paymentIds = ticketToPayments.getOrDefault(ticketId, new ArrayList<>());
        return paymentIds.stream()
                .map(store::get)
                .collect(Collectors.toList());
    }

    public void update(Payment payment) {
        if (!store.containsKey(payment.getId())) {
            throw new IllegalArgumentException("Payment not found: " + payment.getId());
        }
        store.put(payment.getId(), payment);
        System.out.println("[REPO] Payment updated: " + payment.getId() + " | Status: " + payment.getStatus());
    }
}

File: Interview Problems/Parking Lot/Python Code/repository/payment_repository.py
import uuid
from typing import Dict, List, Optional
from domain.payment import Payment

class PaymentRepository:
    def __init__(self):
        self._store: Dict[uuid.UUID, Payment] = {}
        self._ticket_to_payments: Dict[uuid.UUID, List[uuid.UUID]] = {}

    def save(self, payment: Payment):
        self._store[payment.id] = payment
        if payment.ticket_id not in self._ticket_to_payments:
            self._ticket_to_payments[payment.ticket_id] = []
        self._ticket_to_payments[payment.ticket_id].append(payment.id)
        print(f"[REPO] Payment saved: {payment.id} for ticket: {payment.ticket_id}")

    def find_by_id(self, payment_id: uuid.UUID) -> Optional[Payment]:
        return self._store.get(payment_id)

    def find_by_ticket_id(self, ticket_id: uuid.UUID) -> Optional[Payment]:
        payment_ids = self._ticket_to_payments.get(ticket_id, [])
        if not payment_ids:
            return None
        last_id = payment_ids[-1]
        return self._store.get(last_id)

    def find_all_by_ticket_id(self, ticket_id: uuid.UUID) -> List[Payment]:
        payment_ids = self._ticket_to_payments.get(ticket_id, [])
        return [self._store[pid] for pid in payment_ids if pid in self._store]

    def update(self, payment: Payment):
        if payment.id not in self._store:
            raise ValueError(f"Payment not found: {payment.id}")
        self._store[payment.id] = payment
        print(f"[REPO] Payment updated: {payment.id} | Status: {payment.status.name}")
